crates/sui-graphql-client/src/error.rs

```diff
@@ -0,0 +1,316 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use std::num::ParseIntError;
+use std::num::TryFromIntError;
+
+use cynic::GraphQlError;
+
+use sui_types::types::AddressParseError;
+use sui_types::types::DigestParseError;
+use sui_types::types::TypeParseError;
+
+type BoxError = Box<dyn std::error::Error + Send + Sync>;
+
+/// General error type for the client. It is used to wrap all the possible errors that can occur.
+#[derive(Debug)]
+pub struct Error {
+    inner: Box<ClientError>,
+}
+
+/// Error type for the client. It is split into multiple fields to allow for more granular error
+/// handling. The `source` field is used to store the original error.
+#[derive(Debug)]
+pub struct ClientError {
+    /// Error when the GraphQL server returns an error.
+    pub query_response_error: bool,
+    /// Error when deserialization a value (mostly bcs or base64).
+    pub deserialize_error: bool,
+    /// Error when parsing a value.
+    pub parse_error: bool,
+    /// Graphql server returned an empty response, though it was expected to return a value.
+    pub empty_response_error: bool,
+    /// Error when converting from a type to another.
+    pub conversion_error: bool,
+    /// The original error. Use downcasting to get the original error based on the error type from
+    /// above.
+    pub source: Option<BoxError>,
```

Comment:
All these fields need to be private, since the type itself will be private, and accessors should be added to `Error` to get at these inner fields.